### ISSUE TITLE
[red-knot] support dataclasses

### ISSUE THREAD

**Issue by carljm** [astral-sh/ruff#16651]

**Comment by sharkdp** [comment]
What do we plan here *in a first iteration*? We have no support at all for data classes, but we also don't raise a lot of false positives at the moment(?). Attribute annotations work just like for normal classes:
```py
from dataclasses import dataclass

@dataclass
class C:
    x: int
    y: str | None = None

c = C(1, "foo")  # This call is completely unchecked; so no false positives, but also no true positives

reveal_type(c.x)  # revealed: int
reveal_type(c.y)  # revealed: str | None
```

We currently *do* report false positives if there are generated special methods whose existence we do not model:
```py
@dataclass(order=True)
class C:
    x: int

C(2) > C(1)  # False positive: Operator `>` is not supported for types `C` and `C`
```

**Comment by carljm** [comment]
I think the highest priority item would be correctly synthesizing the `__init__` method, because if/when constructor-call-checking lands, it will cause lots of false positives on dataclasses if we don't have that.

**Comment by AlexWaygood** [comment]
Agreed that synthesising the `__init__` method should definitely be the highest-priority item.

> We currently _do_ report false positives if there are generated special methods whose existence we do not model:
> 
> @dataclass(order=True)
> class C:
>     x: int
> 
> C(2) > C(1)  # False positive: Operator `>` is not supported for types `C` and `C`

I think the second-highest priority item might be synthesising the `__lt__`, `__le__`, `__gt__` and `__ge__` methods for `order=True`, to address this false positive here.

**Comment by sharkdp** [comment]
Initial support for dataclasses has been implemented. See https://github.com/astral-sh/ruff/issues/17675 for a lower-priority follow-up ticket.

**Comment by AlexWaygood** [comment]
I'll close this for now so that it no longer shows up on the alpha milestone of tasks still to be done in the next fortnight. Thanks @sharkdp! 😃
